mautic/mautic
# Sendgrid API

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/242
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR adds an ability to send emails through SendGrid API.

**This PR is based on the https://github.com/mautic/mautic/pull/5384. Make sure it will be merged first.**

#### Steps to test this PR:
##### Setup API
1) Go to https://app.sendgrid.com/settings/api_keys and generate API key. Insert it to the Mautic settings. Try both buttons here (`Test connection` + `Send test email`) to be sure it is working.
2) You can use https://ngrok.com/ to testing if you are on localhost. Install tool, run it and get a public address.
3) Setup webhooks in https://app.sendgrid.com/settings/mail_settings > Event Notification. Setup http://your-domain.com/mailer/sendgrid_api/callback URL. Choose supported events:

![screen shot 2017-12-05 at 15 35 25](https://user-images.githubusercontent.com/1649279/33613360-c3126bc2-d9d4-11e7-9be3-76d016c6697d.png)

Do not forget to save it (on the right side of "Event Notification" row).

##### Test sending emails
1) Create a few emails and try to send it. Contacts should have an email address you have access to.
2) Be sure that you send emails with different contents: only HTML content / only Plain text / both. The plain text only version could be send by `Send test email` button in email configuration.
3) Try a form's send form results too.

##### Test unsubscribtion
1) Create a new contact with email you have access to.
2) Create a new contact in SendGrid with same email address (https://sendgrid.com/marketing_campaigns/ui/contacts/).
3) Create a new group for unsubscribe: https://app.sendgrid.com/suppressions/advanced_suppression_manager
4) Create a new sender in SendGrid (https://sendgrid.com/marketing_campaigns/ui/senders/new). You need to be able to verify this address. This is important only to be able to save a campaign (next step)
5) Create a new campaign in SendGrid: https://sendgrid.com/marketing_campaigns/ui/campaigns?name=&statType=percentage&status=All:
- Select sender and unsubscribe group
- Add an unsubscribe link to content (https://sendgrid.com/docs/Classroom/Basics/Marketing_Campaigns/unsubscribe_groups.html#-Manage-Email-Preferences-Links)
- Save and send campaign.
- Wait few minutes for email to be delivered
6) Click to unsubscribe link and contact should be marked as DNC in Mautic.

#### List deprecations along with the new alternative:
1. 
2. 

#### List backwards compatibility breaks:
1. 
2.
